### DISCUSSION TITLE
Material Point Method in MOOSE (moved from mailing list)

### DISCUSSION THREAD

**Discussion by fdkong** [idaholab/moose#16152] [Q&A General] [answered]
This topic is moved from mailing list 
```

On Tue, Nov 10, 2020 at 10:20 PM Jed Brown <jed@jedbrown.org> wrote:
Fande Kong <fdkong.jd@gmail.com> writes:

> On Tue, Nov 10, 2020 at 3:08 PM Nathan Miller <nathan.a.mill@gmail.com>
> wrote:
>
>> Hi All,
>>
>> I'm collaborating with the ASC PSAAP III program at CU Boulder ( see here
>> <https://www.colorado.edu/engineering/2020/10/01/engineering-leads-new-doe-predictive-science-academic-alliance-program-center-particulate>)
>> and they are using my micromorphic continuum mechanics code (
>> https://github.com/lanl/tardigrade ). As a part of the project, they
>> would like to try and use the material point method to solve the
>> micromorphic balance equations. Jed Brown, who is also collaborating on the
>> project, mentioned that PETSc's DMSwarm capabilities might be able to do
>> this but that it may require modification to libMesh or for MOOSE to
>> support some additional PETSc features that he wasn't sure you did.
>>
>
> Yes, we need to refactor DM a bit to bring DMSwarm in, and do some
> transfers between MOOSE MESH and PETSc DM. It is interesting for me.
> Unfortunately, we do not have any funded project that has such
> requirements right now. That being said, we might not get that really soon.
>
> Outside contributions are always welcome : -)

Thanks, Fande. We may be able to drive this forward under the PSAAP, but would
 like some guidance from the MOOSE team about how you would like to see it
 integrated. If you or other MOOSE developers have thoughts on that, perhaps
 we could discuss be email or a short screen-share meeting. This is a bit preliminary
 in that Nathan and myself are the only members of our current team who have
 experience with MOOSE, but I don't think either of us have free cycles to start
 implementing this right away. (We are hiring, if anyone is reading this and interested...)

The PSAAP center's needs include MPM for micromorphic mechanics (with Tardigrade)
 and MPM for grain-resolving simulation that is much more performance sensitive and
 will likely not use MOOSE. The project just started and strategy for the latter
 component is still in flux, but there's a good chance it will use DMSwarm
 (and upstream anything relevant).
```

**Comment by fdkong** [comment, marked as answer]
@jedbrown want to introduce DMSwarm to MOOSE for MPM.

@idaholab/moose-team Any thoughts on how to integrate this capability?

**Reply by lindsayad** [reply]
Let's make sure to pull in @nathan-miller-LANL (I think I got this tag right). I think I could give a more educated answer if I was a little more educated on `DMSwarm`. I am always intrigued by a potential increase in integration between PETSc and libMesh/MOOSE. I think it would be a great idea to sit down and do the screen-share meeting that @jedbrown bought up. We would make sure to dial @roystgnr into that call as well

**Reply by nathan-miller-LANL** [reply]
Yep, that's me. Let me know when you all want to try and to the web-meeting.

**Reply by fdkong** [reply]
Nathan, go ahead to schedule a meeting next week or so. 

It might be very really helpful if you could have a background introduction of this problem because we might not be familiar with this topic.

**Reply by mcacace** [reply]
Dear all,
I followed with interest this discussion. I will have a post doctoral position starting (hopefully) soon where the candidate, among other things, would be asked to (at least try) to integrate Particle in Cell (PIC) capabilities to one of our moose based app, possibly based on DMSwarm from Petsc. This said, even if for a different applications than what discussed above, we would also be interested in any progress made in this field, and will also be willing to actively do some coding from our part. My comment  was meant on an informative level, but I would be glad to be informed if possible of any progress and/or decision about it. Thanks and sorry for stepping in.
Mauro

**Reply by nathan-miller-LANL** [reply]
@fdkong, okay. I'll try and coordinate that.

@mcacace I'm glad to hear others are interested. Note that the method that we would want to implement MPM for would be a higher order continuum theory which includes as a subset classical (i.e. Cauchy) continuum mechanics. I would hope that what we would do would be at least tangentially applicable to your problem.

**Comment by wjin33** [comment]
I am working on granular flow physics, and would like to see the development of MPM in MOOSE.

**Comment by GiudGiud** [comment]
@wjin33 @loganharbour @friedmud and I have taken a look at mpm in moose if people want to join in on the effort

**Reply by mcacace** [reply]
@GiudGiud et al: this is great news! As written in my previous post, we have some applications, for which usage of mpm (was considering more PIC) methods based on DMSwarm Petsc capabilities would be an asset. Would be interesting to learn more about the status and try to help the development. 
Mauro
